### PR TITLE
Match engine

### PR DESCRIPTION
**Pull request by iZMC**

### CODE REVIEW

**Comment by iZMC** [comment]
## TODO
1. match end condition  refactor
2. const (PRIVATE_KEY_PATH..) adjust
3. change order status

File: src/modules/orders/order.repository.ts [hunk 1]
```diff
@@ -47,14 +48,32 @@ class OrderRepository extends Repository<Order> {
...
           output: JSON.stringify(cell.output),
           type: cell.type === '00' ? OrderType.Bid : OrderType.Ask,
           price: cell.price.toString(16).padStart(16, '0'),
         })
+
+  private isInvalidOrderCell(cell: ReturnType<typeof parseOrderCell>) {
+    const smallestCapacity: bigint =
+      (cell.orderAmount * cell.price * BigInt(1003)) / BigInt(1000) / BigInt(10 ** 13) + BigInt(17900000000)
+    const biggestCapacityOrderAmount: bigint =
+      ((cell.sudtAmount - (cell.sudtAmount * BigInt(3)) / BigInt(1000)) * cell.price) / BigInt(10 ** 10)
```
Comment (dex-dev-outlook) [inline src/modules/orders/order.repository.ts:64]: Don't use magic number `1003`, `1000`, `10 ** 13`, `17900000000`, `3`, `10 ** 10`

File: src/modules/orders/order.repository.ts [hunk 2]
```diff
@@ -47,14 +48,32 @@ class OrderRepository extends Repository<Order> {
...
+
+    return (
+      cell.orderAmount === BigInt(0) ||
```
Comment (dex-dev-outlook) [inline src/modules/orders/order.repository.ts:67]: `cell.orderAmount === BigInt(0)` can be changed to `!cell.orderAmount` for concision.

And don't too much chaining `||`, return `false` in each falsy branch.

File: src/modules/orders/index.ts [hunk 3]
```diff
@@ -5,21 +5,97 @@ import OrderRepository from './order.repository'
...
+  SECP256K1_CODE_HASH,
+  SECP256K1_TX_HASH,
+  ORDERBOOK_TX_HASH,
+} from '../../utils'
 import { Deal, DealStatus } from './deal.entity'
-import { logger, parseOrderCell } from '../../utils'
+import fs from 'fs'
+import CKB from '@nervosnetwork/ckb-sdk-core'
+import { getTransactionSize } from '@nervosnetwork/ckb-sdk-utils/lib/sizes'
+import { Indexer, CellCollector } from '@ckb-lumos/indexer'
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:24]: Imports from `node_modules` should be placed above those from modules of the project.

File: src/modules/orders/index.ts [hunk 4]
```diff
@@ -5,21 +5,97 @@ import OrderRepository from './order.repository'
...
+  fee: bigint = BigInt(3)
+  feeRatio: bigint = BigInt(1_000)
+  shannonsRatio: bigint = BigInt(100_000_000)
+  priceRatio: bigint = BigInt(10_000_000_000)
+
+  /*
+  This function is the entry for match engine.
+  Our workflow is get sorted order list from sync module.Then match them, generate raw
+  transaction, sign and send it.
+  In this function ewe do some check to ensure our matcher can successfully work.
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:53]: Is `ewe` a typo?

File: src/modules/orders/index.ts [hunk 5]
```diff
@@ -5,21 +5,97 @@ import OrderRepository from './order.repository'
...
+  Such as set private key; check order list is not empty; check deal maker have live cells etc.
+  Other sync order filter conditions are define in orders/order.repository.ts toCell function.
+  */
+  public async prepareMatch(indexer: Indexer, sudt_type_args: string) {
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:57]: Use `camelCase`

File: src/modules/orders/index.ts [hunk 6]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+    if (config.keyFile === null) {
+      return {
+        privateKey: null,
+        dealMakerLock: undefined,
+      }
+    }
+
+    const privateKey = fs.readFileSync(config.keyFile, 'utf-8').trim()
+    const publicKey = ckb.utils.privateKeyToPublicKey(privateKey)
+    const hexPublicKey = `0x${ckb.utils.blake160(publicKey, 'hex')}`
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:168]: `hexPublicKey` should be renamed to `publicKeyHash`

File: src/modules/orders/index.ts [hunk 7]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+  Match engine is design for meet user need, and we have order amount in cell data to define this need.
+  So for ask order, you need to buy enough CKB amount.We don't care it's SUDT amount left.
+  For bid order, you need to buy enough SUDT amount. We don't care it's ckb amount left.
+  For match there have two condition: 1. full matched 2. partially matched.
+  Full matched means two orders need both can be meet.
+  In other words, ask order's need ckb amount is equal to bid order can pay for it's SUDT value.
+  Partially matched means only one direction's need can be meet.So one order will be deal done,
+  another will have left.
+  */
+  private startMatchAndReturnOutputs(bidOrderList: Array<OrderDto>, askOrderList: Array<OrderDto>): any {
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:192]: Don't use type `any`

File: src/modules/orders/index.ts [hunk 8]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+        return []
+      }
+    } else {
+      // We calculate deal price by a simmplest strategy, use average price.After all calculation we use deal,so forget order price.
+      const dealPrice: bigint = (askCapacityPrice + bidCapacityPrice) / BigInt('2')
+
+      const bidSudtAmount: bigint = parseOrderData(bidOrderOutput.data).sudtAmount
+      const bidSudtOrderAmount: bigint = parseOrderData(bidOrderOutput.data).orderAmount
+      const bidActualSpendCapacityAmount: bigint = (dealPrice * bidSudtOrderAmount) / this.shannonsRatio
+      const bidOriginalCapacityAmount: bigint = BigInt(bidOrderOutput.capacity)
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:244]: They can be organized in a object named `bid` so the reference can be more clear, as `bid.SudtAmount`, bid.SudtOrderAmount`

File: src/modules/orders/index.ts [hunk 9]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+
+  private generateNewOutput(
+    originalOrderCell: OrderDto,
+    capacityAndSudt: { capacity: string; data: string },
+    originalScript: { lock: object; type: object },
+  ) {
+    const newOutputCell: { capacity: string; lock: object; type: object } = {
+      ...{ capacity: capacityAndSudt.capacity },
+      ...originalScript,
+      ...{ data: capacityAndSudt.data },
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:454]: `capacity` and `data` can be set directly.

File: src/modules/orders/index.ts [hunk 10]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+    }
+    return { ...originalOrderCell, ...{ output: JSON.stringify(newOutputCell) }, ...{ part: true } }
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:456]: `output` and `part` can be set directly

File: src/modules/orders/index.ts [hunk 11]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+    return BigInt(getTransactionSize(rawTx)) * this.feeRatio
+  }
+
+  private subMinerFeeAndUpdateOutputs(
+    rawTx: CKBComponents.RawTransactionToSign,
+    minerFee: bigint,
+  ): CKBComponents.RawTransactionToSign {
+    const dealMakerOutput: CKBComponents.Cell = rawTx.outputs[0]
+    const dealMakerCapacity = '0x' + (BigInt(rawTx.outputs[0].capacity) - minerFee).toString(16)
+    const newDealMakerOutput: CKBComponents.Cell = { ...dealMakerOutput, ...{ capacity: dealMakerCapacity } }
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:540]: `capacity` can be set directly.

File: src/modules/orders/index.ts [hunk 12]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+    this.outputsCells.shift()
+    this.outputsCells.unshift(newDealMakerOutput)
+    return { ...rawTx, ...{ outputs: this.outputsCells } }
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:543]: `outputs` can be set directly

File: src/modules/orders/index.ts [hunk 13]
```diff
@@ -67,13 +143,479 @@ class OrdersService {
...
+    lock: CKBComponents.Script,
+    privateKey: string,
+    deal: { txHash: string; status: number; orderIds: string; fee: string; tokenId: string },
+  ) {
+    try {
+      const response = await signAndSendTransaction(rawTransaction, privateKey, lock)
+      deal.txHash = response
+
+      console.info('==============================')
+      console.info(response)
```
Comment (dex-dev-outlook) [inline src/modules/orders/index.ts:601]: Use `logger` to print info with prefix
Ref: https://github.com/glias/deal-maker-cli/blob/develop/src/modules/orders/index.ts#L15-L17

**Review by Keith-CY — APPROVED** [review]

**Review by dex-dev-outlook — APPROVED** [review]

**Pull request merged**